Python/Product/PythonTools/PythonTools/Debugger/DebugAdapter/DebugAdapterAttachProcess.cs

```diff
@@ -0,0 +1,209 @@
+                debugPyProcess.ErrorDataReceived += DebugPyProcess_OutputDataReceived;
+            }
+        }
+
+        private void DebugPyProcess_OutputDataReceived(object sender, DataReceivedEventArgs e) {
+            Debug.WriteLine($"DebugPy output: {e.Data}");
+        }
+
+        private void DebugPyProcess_Exited(object sender, EventArgs e) {
+            Debug.WriteLine($"DebugPy exited:");
+            Exited?.Invoke(this, EventArgs.Empty);
+        }
+
+        public static ITargetHostProcess LocalAttach(DebugAttachInfo debugAttachInfo, IAdapterLaunchInfo launchInfo, ITargetHostInterop targetInterop) {
+            // For the local case, we need to start the process ourselves
+            AD_PROCESS_ID adProcessId = new AD_PROCESS_ID();
+            adProcessId.ProcessIdType = (uint)enum_AD_PROCESS_ID.AD_PROCESS_ID_SYSTEM;
+            adProcessId.dwProcessId = (uint)launchInfo.AttachProcessId;
+            launchInfo.DebugPort.GetProcess(adProcessId, out var process);
+            process.GetName(VisualStudio.Debugger.Interop.enum_GETNAME_TYPE.GN_FILENAME, out var processName);
+            var debugPyServerDirectory = Path.GetDirectoryName(PythonToolsInstallPath.GetFile("debugpy\\__init__.py"));
+            var port = GetAvailablePort();
+
+            // Start the process using the --listen --pid arguments
+            var debugPyProcess = new Process();
+            debugPyProcess.StartInfo.FileName = processName;
+            debugPyProcess.StartInfo.Arguments = $"\"{debugPyServerDirectory}\" --listen {port} --pid {launchInfo.AttachProcessId}";
+            debugPyProcess.StartInfo.RedirectStandardError = true;
+            debugPyProcess.StartInfo.RedirectStandardInput = true;
+            debugPyProcess.StartInfo.RedirectStandardOutput = true;
+            debugPyProcess.StartInfo.UseShellExecute = false;
+            debugPyProcess.StartInfo.CreateNoWindow = true;
+            debugPyProcess.EnableRaisingEvents = true;
+            var attachedProcess = new DebugAdapterAttachProcess(debugPyProcess);
+            debugPyProcess.Start();
+
+            // Connect to this port
+            return attachedProcess.ConnectSocket(new System.Uri($"http://localhost:{port}"), debugPyProcess) ? attachedProcess : null;
+        }
+
+        public static ITargetHostProcess RemoteAttach(DebugAttachInfo debugAttachInfo) {
+            var attachedProcess = new DebugAdapterAttachProcess();
+            return attachedProcess.ConnectSocket(debugAttachInfo.RemoteUri) ? attachedProcess : null;
+        }
+
+        /// <summary>
+        /// Gets an available port on the loopback interface.
+        /// </summary>
+        private static int GetAvailablePort() {
+            TcpListener l = new TcpListener(IPAddress.Loopback, 0);
+            l.Start();
+            int port = ((IPEndPoint)l.LocalEndpoint).Port;
+            l.Stop();
+            return port;
+        }
```

Comment:
This is not a reliable way to do this, because the port can be taken by some other process as soon as you close the listener socket.

It would be better to implement this using `debugpy --connect` to invert the direction, so that you can just keep the socket open on VS side, and meanwhile tell debugpy where to connect.

Alternatively, when using `--listen`, you can tell debugpy to listen on :0, and get the actual port number by setting the environment variable `DEBUGPY_ADAPTER_ENDPOINTS` to some temp filename. When adapter starts listening for client connections, it'll write a JSON snippet into that file which has the actual host & port being listened on. This is still not ideal, though, because you have to poll for that file in a loop - it was really added mostly for testing purposes where that's okay.